[moveonly] Extract RescanWallet to handle a simple rescan

Empact opened the pull request:
Where the outcome does not depend on the result, apart from a simple
success check.

Empact committed:
[moveonly] Extract RescanWallet to handle a simple rescan

Where the outcome does not depend on the result, apart from a simple
success check.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -935,7 +933,6 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     void BlockConnected(const std::shared_ptr<const CBlock>& pblock, const CBlockIndex *pindex, const std::vector<CTransactionRef>& vtxConflicted) override;
     void BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) override;
     bool AddToWalletIfInvolvingMe(const CTransactionRef& tx, const CBlockIndex* pIndex, int posInBlock, bool fUpdate) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
-    int64_t RescanFromTime(int64_t startTime, const WalletRescanReserver& reserver, bool update);
```
Is there any reason why not just leave this here? It could very well be used by the gui or some other interface to the wallet in the future and had to be moved back. Generally, I'd prefer if code was only moved when necessary, since unnecessary code movement/refactoring makes backports easier to mess up and consumes scarce review resources.

Empact replied:
This move reduces the surface area of CWallet, which is pretty sprawling at the moment (wallet.cpp is ~4500 lines, the 3rd-largest *.cpp file, and the largest outside `/qt/`), and localizes this code to one implementation file from being more broadly available (wallet.h is included into 26 files).

Grant the concerns you note are serious, and we should draw the line somewhere, but if the codebase is to be in play for a long time, it needs to evolve and progress. Moving incrementally toward minimizing scope of code, dependencies, etc. is attractive from that perspective as it should have payoffs in reducing the overall complexity of the codebase, which should ease changes on an ongoing and cumulative basis.

Other approaches to making better use of scarce review resources come to mind:
* I've thought of implementing a web site to helps rank PRs by ACKs and other indications of importance. Current solutions like: https://github.com/bitcoin/bitcoin/projects/8 are pretty limited and not decentralized and ACKs are not currently well-exposed across PRs in aggregate.
* We could add a feature to output the changes between force pushes to the same PR, so that reviewers could review just the incremental subset, and the historical changes are recorded.
* We could potentially examine and report changes to the build output to give reviewers a sense of the consequence to the output to the binaries.

Seems like these are the sorts of things that will help the project to scale and accelerate.

MarcoFalke replied:
> I've thought of implementing a web site to helps rank PRs by ACKs and other indications of importance. Current solutions like: https://github.com/bitcoin/bitcoin/projects/8 are pretty limited and not decentralized and ACKs are not currently well-exposed across PRs in aggregate.

Are you aware of https://bitcoinacks.com/?desc=1&flt3_label_not_contains=Needs+rebase&flt0_merged_empty=1&flt4_ci_contains=Success&flt1_closed_empty=1&sort=7&flt2_mergeable_contains=Mergeable

Empact replied:
Nice, hadn't see that. 👍

Empact commented:
This refactor raises the following questions:
* Why does rpc `importmulti` call `pwallet->ReacceptWalletTransactions` before aborting on `pwallet->IsAbortingRescan()`. Should it flip that ordering like the others do?
* Why do 3 of the 5 calls call `pwallet->ReacceptWalletTransactions`, while rpc `importwallet` calls `pwallet->MarkDirty()` and rpc `importprivkey` calls neither. This seems like a possible bug in rpc `importprivkey`.

Empact commented:
Updated to minimize `git diff --color-moved=dimmed_zebra --minimal head^^`

sipa commented:
I don't like this. Code that interacts with the wallet should be in the wallet, not in the RPC code. I'm aware that there is an enormous amount of internal logic already in the RPC code, but we should be moving stuff in the other direction.

MarcoFalke commented:
utACK second commit, if the first one is dropped.

Empact commented:
Dropped the first commit, and added handling of the ShutdownRequested rescan abort condition:
https://github.com/bitcoin/bitcoin/blob/b25a4c2284babdf1e8cf0ec3b1402200dd25f33f/src/wallet/wallet.cpp#L1759-L1763

promag commented:
Kind of NACK the current approach. If you want to touch this consider improving `CWallet::ScanForWalletTransactions` signature, like:
```cpp
enum WalletScanError
{
    NONE = 0, // when ScanForWalletTransactions return is true
    ABORTED,
    SHUTDOWN_REQUEST,
};
struct WalletScanResult {
    WalletScanError error;
    std::string message;
    // other outputs
};
bool CWallet::ScanForWalletTransactions(WalletScanResult& result, ...);
```
This way the callers don't have to check what caused the error.

Empact commented:
@promag I did something like that here #13076, and while I think the result is ok, I've been thinking more of updating it to return bool as prompted by jimpo in #13582:
https://github.com/bitcoin/bitcoin/pull/13582#issuecomment-401889114

Have that follow-up in mind.

Empact commented:
I'll split these up for separate consideration.

promag commented:
@Empact my suggestion is in line with @jimpo suggestion.

promag commented:
utACK 3fe836b.

Side note, not sure if this should be tagged moveonly.

MarcoFalke commented:
utACK 3fe836b78d504942e8850b607453886969f57e27

laanwj commented on src/wallet/wallet.h:
```diff
@@ -62,8 +62,6 @@ static const bool DEFAULT_WALLET_RBF = false;
 static const bool DEFAULT_WALLETBROADCAST = true;
 static const bool DEFAULT_DISABLE_WALLET = false;
 
-static const int64_t TIMESTAMP_MIN = 0;
```
good to move this locally, it's a really aspecific constant name
